### PR TITLE
JAMES-2257 Allow Message::isDraft property modification via JMAP

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature [hunk 1]
```diff
@@ -180,6 +180,11 @@ Feature: SetMessages method on shared folders
     When "bob@domain.tld" moves "mBob" to user mailbox "Drafts"
     Then message "mBob" is updated
 
+  Scenario: A user can update $Draft keyword using isDraft property
+    Given "bob@domain.tld" has a mailbox "Drafts"
+    When "bob@domain.tld" marks the message "mBob" as daft
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: s/daft/draft

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java [hunk 2]
```diff
@@ -79,6 +80,11 @@ public Builder isAnswered(Boolean isAnswered) {
             return this;
         }
 
+        public Builder isDraft(Boolean isAnswered) {
+            this.isDraft = Optional.of(isAnswered);
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java]: s/isAnswered/isDraft

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/OldKeywordTest.java [hunk 3]
```diff
@@ -79,8 +91,8 @@ public void applyStateShouldReturnSeenWhenPatchSetsSeenOnSeenMessage() {
         Optional<Boolean> isAnswered = Optional.empty();
         Optional<Boolean> isDraft = Optional.empty();
         Optional<Boolean> isForwarded = Optional.empty();
-        OldKeyword testee = new OldKeyword(isUnread, isFlagged, isAnswered, isDraft, isForwarded);
-        
-        assertThat(testee.applyToState(new Flags(Flag.SEEN))).isEqualTo(new Flags(Flag.SEEN));
+        Optional<OldKeyword> testee = OldKeyword.computeOldKeywords(isUnread, isFlagged, isAnswered, isDraft, isForwarded);
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/OldKeywordTest.java]: s/computeOldKeywords/compute/ ? or s/computeOldKeywords/of/ ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java [hunk 4]
```diff
@@ -204,20 +204,22 @@ public void setFlags(String username, List<String> keywords, String message) thr
 
     @When("^\"([^\"]*)\" marks the message \"([^\"]*)\" as flagged$")
     public void flag(String username, String message) throws Throwable {
-        MessageId messageId = messageIdStepdefs.getMessageId(message);
+        userStepdefs.execWithUser(username, () -> {
```
Comment (strokyl) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java]: I was currently fixing that...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/OldKeyword.java [hunk 5]
```diff
@@ -28,6 +28,14 @@
 import com.google.common.base.Objects;
 
 public class OldKeyword {
+    public static Optional<OldKeyword> computeOldKeywords(Optional<Boolean> isUnread, Optional<Boolean> isFlagged, Optional<Boolean> isAnswered,
```
Comment (strokyl) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/OldKeyword.java]: It's going to be so easy to miss up in the order

Comment (rouazana) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/OldKeyword.java]: yes but I expect this code to be deleted in some weeks, so no need to invest more on this (more than good unit test to ensure this error does not happen)

Comment (strokyl) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/OldKeyword.java]: I propose you a solution with something looking like a builder.
I might have rush a little bit on naming/testing.
If you like it feel free to improve it

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java (hunk 2 above)
Comment (strokyl) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java]: done in my last commit

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature (hunk 1 above)
Comment (strokyl) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SetMessages.feature]: done

**Comment by chibenwa** [comment]
:+1:

**Comment by chibenwa** [comment]
Re-reading this, I was not happy with the overall code quality of the **OldKeywords** overall feature.

I don't know where but we lost isolation of that design choice in recent refactorings. I will open a PR that re-enforce this.

**Pull request closed without merging**